src/hotspot/share/gc/parallel/psOldGen.cpp

```
@@ -190,7 +190,7 @@ bool PSOldGen::expand(size_t bytes) {
   assert(bytes > 0, "precondition");
 #endif
   const size_t alignment = virtual_space()->alignment();
-  size_t aligned_bytes  = align_up(bytes, alignment);
+  size_t aligned_bytes = can_align_up(bytes, alignment) ? align_up(bytes, alignment) : 0;
```

Why doesn't the previous revision using early-return + min2 work?

---

> try to expand instead of aborting.

Did the previous version abort? I thought the logic is essentially:

```
if (remaining == 0) {
  return;
}

aligned_bytes = align_up(min2(bytes, remaining), alignment).
```

The below `if (aligned_bytes == 0) { ` should not be reachable any more, since align-up-overflow can't occur.

---

I meant aborting in the sense that we abort the expand operation by returning early instead of continuing.

If we don't that, the `if (aligned_bytes == 0) {` should be reachable with `can_align_up` if the align were to fail and we set aligned_bytes to 0. The `remaining == 0` would then be caught later in the function inside `expand_by(aligned_bytes)`.

I'll go ahead and revert to the other version, and remove the `if (aligned_bytes == 0) {` block as well.